### PR TITLE
Increase Processing Speed of Traffic Flow Objects using multi-processing

### PR DESCRIPTION
**Pull request by RonGonz-illumio**
Could be considered an alternative to removing validate
* decides whether to use single process or multipe processes depending on number of flows
* displays tqdm progress bar when using multiprocessing
* uses chunk sizes to minimize number of rpc calls to transfer data

pce.py
- Added imports for concurrent.futures, multiprocessing, and tqdm.
- Refactored the method `_get_traffic_flows()` to use multiprocessing for improved performance when the number of flows is greater than 10,000.
- Added a static method `_transform_traffic_flows()` to handle the transformation of raw flow data into TrafficFlow objects.

requirements.txt:
- Updated the version of pytest to `7.2.2`.
- Added tqdm (`4.66.1`) and urllib3 (`1.26.15`) to the requirements.

cProfile results for single process (original list comprehension) method:
```
Traffic Search Returned: 74 flows, using single process...
         329733963 function calls (310837371 primitive calls) in 132.544 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        4    0.025    0.006  133.438   33.360 pce.py:945(<listcomp>)
1129211/18353    5.047    0.000  133.413    0.007 jsonutils.py:111(from_json)
1129211/264934    0.446    0.000  101.614    0.000 jsonutils.py:46(__post_init__)
1129211/264934    6.207    0.000  101.378    0.000 jsonutils.py:141(_decode_complex_types)
1129211/264934    4.958    0.000  101.278    0.000 jsonutils.py:49(_validate)
9865408/2796931    3.067    0.000   98.476    0.000 jsonutils.py:56(_validate_field)
9791996/2796931    3.188    0.000   98.080    0.000 jsonutils.py:147(_decode_field)
  1129211    0.572    0.000   89.312    0.000 inspect.py:3276(signature)
  1129211    0.843    0.000   88.740    0.000 inspect.py:3022(from_callable)
2258422/1129211   11.030    0.000   87.896    0.000 inspect.py:2425(_signature_from_callable)
    18353    0.044    0.000   65.822    0.004 trafficanalysis.py:286(_validate)
  1129211   15.194    0.000   50.671    0.000 inspect.py:2330(_signature_from_function)
   111918    0.131    0.000   40.198    0.000 {built-in method builtins.all}
   537074    0.266    0.000   40.067    0.000 jsonutils.py:87(<genexpr>)
   537074    0.241    0.000   38.407    0.000 jsonutils.py:158(<genexpr>)
    77016    0.117    0.000   23.803    0.000 workload.py:167(_validate)
    77016    0.102    0.000   23.565    0.000 workload.py:174(_decode_complex_types)
 10496051   13.643    0.000   22.317    0.000 inspect.py:2682(__init__)
  2258422   12.898    0.000   19.456    0.000 inspect.py:2968(__init__)
  1129211    2.620    0.000   15.347    0.000 inspect.py:2045(_signature_bound_method)
  1129211    0.898    0.000   12.172    0.000 inspect.py:3038(replace)
  2258422    4.184    0.000    7.649    0.000 dataclasses.py:1223(fields)
 43644123    4.532    0.000    5.844    0.000 {built-in method builtins.isinstance}
 10496051    4.056    0.000    5.777    0.000 enum.py:689(__call__)
 11625262    3.262    0.000    4.263    0.000 inspect.py:3017(<genexpr>)
  2258422    2.111    0.000    3.783    0.000 inspect.py:735(unwrap)
 20992102    2.968    0.000    2.968    0.000 dataclasses.py:1238(<genexpr>)
 27767368    2.585    0.000    2.585    0.000 {built-in method builtins.getattr}
  1129211    1.667    0.000    2.271    0.000 inspect.py:167(get_annotations)
  1129211    1.043    0.000    2.254    0.000 inspect.py:2071(_signature_is_builtin)
  3387633    1.361    0.000    2.149    0.000 inspect.py:1953(_signature_get_user_defined_method)
  2938422    0.962    0.000    2.119    0.000 {built-in method builtins.issubclass}
```

### CODE REVIEW

**Comment by RonGonz-illumio** [comment]
take a look at the new commit which should address all the above issues

File: illumio/pce.py [hunk 1]
```diff
@@ -940,17 +943,30 @@ def get_traffic_flows_async(self, query_name: str, traffic_query: TrafficQuery,
...
-
-            collection_href = self._async_poll(location)
-
-            response = self.get(collection_href)
-            response.raise_for_status()
-            return [TrafficFlow.from_json(flow) for flow in response.json()]
+            print(f"Transforming Json data into TrafficFlow Objects...")
+            raw_flow_data = response.json()
+            rfd_len = len(raw_flow_data)
+            if rfd_len == 0:
```
[the remark this answers is not in the document]
Comment (dsommerville-illumio) [reply illumio/pce.py:951]: still don't think this if is needed. 

```
                if rfd_len > 10000:
                    cpu_count = multiprocessing.cpu_count()
                    chunksize = int(rfd_len / (10 * cpu_count))
                    with concurrent.futures.ProcessPoolExecutor(max_workers=cpu_count/2) as executor:
                        return list(executor.map(TrafficFlow.from_json, raw_flow_data, chunksize=chunksize))
                else:
                    results = [TrafficFlow.from_json(flow) for flow in raw_flow_data]
```

should cover the 0 length case just fine

File: requirements.txt [hunk 2]
```diff
@@ -4,3 +4,5 @@ pytest
 pytest-cov
 requests>=2.26.0
 requests-mock
+
+tqdm~=4.66.1
```
Comment (dsommerville-illumio) [inline requirements.txt:8]: can you take out tqdm from here?

**Review by dsommerville-illumio — CHANGES_REQUESTED** [review]
Cool, this looks better. Couple nits below. One thing I'm wondering is if the mp should be a wrapper classmethod in `JsonObject` - pass in the `response.json()` results and handle extraction in one place so we don't need to rewrite it if we want to reuse for other types.

**Comment by RonGonz-illumio** [comment]
> Cool, this looks better. Couple nits below. One thing I'm wondering is if the mp should be a wrapper classmethod in `JsonObject` - pass in the `response.json()` results and handle extraction in one place so we don't need to rewrite it if we want to reuse for other types.

I implemented this instead, but I think from_json should really do the determining of whether to use normal method or mp method.. however that leaves less choice for the user to determine which one he wants to use.

**Comment by dsommerville-illumio** [comment]
> > Cool, this looks better. Couple nits below. One thing I'm wondering is if the mp should be a wrapper classmethod in `JsonObject` - pass in the `response.json()` results and handle extraction in one place so we don't need to rewrite it if we want to reuse for other types.
> 
> I implemented this instead, but I think from_json should really do the determining of whether to use normal method or mp method.. however that leaves less choice for the user to determine which one he wants to use.

the difference imo is that `from_json` should always be a single object - another function like `convert_results` or something that can be passed the collection and then determine whether or not to use mp is what I'm thinking of. then we can just replace all the calls to e.g. `return [ObjectType.from_json(o) for o in response.json()]` with `return ObjectType.convert_results(response.json())`.

**Comment by RonGonz-illumio** [comment]
> > I implemented this instead, but I think from_json should really do the determining of whether to use normal method or mp method.. however that leaves less choice for the user to determine which one he wants to use.
> 
> the difference imo is that `from_json` should always be a single object - another function like `convert_results` or something that can be passed the collection and then determine whether or not to use mp is what I'm thinking of. then we can just replace all the calls to e.g. `return [ObjectType.from_json(o) for o in response.json()]` with `return ObjectType.convert_results(response.json())`.

that sounds cool - I'm going to leave that up to you, not sure what you want to do now with this pull request..

**Comment by RonGonz-illumio** [comment]
I have read the CLA Document and I hereby sign the CLA

File: illumio/pce.py [hunk 3]
```diff
@@ -24,10 +24,11 @@
 License:
     Apache2, see LICENSE for more details.
 """
+import concurrent.futures
```
Comment (dsommerville-illumio) [inline illumio/pce.py:27]: these are no longer used here, right?

Comment (RonGonz-illumio) [reply illumio/pce.py]: yah - killed them too

**Review by dsommerville-illumio — APPROVED** [review]
LGTM, thanks!

I'll try out that tweak generalizing your mp function for use in other GET calls in another branch.

**Pull request merged**